### DISCUSSION TITLE
Create authentication flow with mandatory 2FA with OTP or WebAuthn

### DISCUSSION THREAD

**Discussion by claudioweiler** [keycloak/keycloak#14988] [Q&A] [answered]
Hi,

KeyCloak comes with default browser authentication flow with OTP 2FA `Conditional` flow configured  (`Forms - Auth-otp-form - Conditional`). If this flow is changed to `Required`, then OTP will be mandatory, and user must configure one on login if he do not have one configured yet.

How to add an additional WebAuthn Authenticator execution as alternative to OTP, but continue with mandatory 2FA?

I tried to add WebAuthn Authenticator in same level of OTP, and changed both to "Alternative", this works since user has one of them configured, but to new users the authentication fails with log `KC-SERVICES0013: Failed authentication: org.keycloak.authentication.AuthenticationFlowException`.

![image](https://user-images.githubusercontent.com/15875077/196530992-0f934cb3-2c2d-46fc-89aa-6e00e03d1d41.png)

I'm trying to understand the execution `Condition - User Configured`, I'm think that the gotcha lives there. But all test that I make fails miserably: or authentication fails, or OTP is always required without the link to use another method.

**Comment by claudioweiler** [comment, marked as answer]
Hi!

I think I found a way to this. Added this flow below `Username Password Form`:
```
- Browser 2FA: required
  - Authenticate: alternative
    - Condition - User Configured: required
	- OTP Form: alternative
	- WebaAthn Authentication: alternative
  - Register: alternative
    - OTP Form: required
```

Some tests, looks OK.

If someone can points any problem with this, or have a better way, please share.

**Reply by darius-m** [reply]
I think that this document (https://github.com/keycloak/keycloak-community/blob/main/design/multi-factor-admin-and-step-up.md) may provide with a better answer, but I have not read it in its entirety.

From what I remember reading a while back, the logic in Keycloak is that the `Required` executions in a subflow take precedence and completely disable all `Alternative` executions and subflows. In the first example that you have given, only the condition, which was set to `Required` is checked, and the credentials are never checked. This is because Keycloak sees that the execution entries are set as `Alternative`, which is equivalent to `Disabled` because of the `Required` condition. I'm not how you can add a condition to check if any of the 2FA options is enabled or not.

The ability to choose from combined (OTP, WebAuthn, etc.) 2FA options was added after Keycloak 15 (in either 16 or 17). The preexisting system was extended such as if multiple 2FA options are set as alternatives in a subflow, the menu to choose from them is displayed on the user side. However, the `Required` / `Alternatives` requirement relation has been kept the same. From what I know, the flow that you have presented as the answer above is the expected approach.

Before this change was made, the authentication either immediately failed in a similar manner as the one you have described, or you would get the first option, without the ability to ever choose the second (e.g. if you had OTP and WebAuthn as the alternatives, you would always get the prompt for OTP, but could not select WebAuthn). I cannot remember which of them it was, however.

**Reply by claudioweiler** [reply]
Hi @darius-m. Thanks to pointing this document, it brings some lights.

> In the first example that you have given, only the condition, which was set to Required is checked, and the credentials are never checked

They are checked, indeed. But it works only if user is (already) configured, if user isn't configured then authentication fails. If only one MFA method is used and is marked with required, then, if user isn't configured, it requires the user to configure the MFA method.

>  I'm not how you can add a condition to check if any of the 2FA options is enabled or not.

This is job to the `Condition - User Configured` execution. I'm trying to understand better how this works, I'm not confident with his behavior yet.

**Reply by darius-m** [reply]
I forgot how it was. I believe that you are correct, and alternatives would never ask users to configure the second factor if none of them was configured. In the example above (the correct one), are users not asked to configure a second factor token on their first login? If they are, I would assume that that the flows work as expected, and users can go to the user account console and configure the tokens as they wish after logging in. 

If not, there was one way to "brute force" it, by having the OTP and WebAuthn tokens as `Required` steps in `Alternative` subflows (basically turn the executions above in subflows that only contain one `Required` execution). But this may very well lead to the scenario I was talking about before, where Keycloak would choose the first alternative and ask you for OTP, without giving you the option to select the alternative.

**Reply by claudioweiler** [reply]
> In the example above (the correct one), are users not asked to configure a second factor token on their first login? If they are, I would assume that that the flows work as expected, and users can go to the user account console and configure the tokens as they wish after logging in.

They are asked. In this case, they are asked to configure OTP in `Register` flow. I can't make a flow where user can choose type of MFA to configure on first login (mandatory). After successful login, then users can manage they MFA.

**Reply by darius-m** [reply]
It seems that you can somewhat force users to register an OTP option on their first login using this flow:
```
- Browser with 2FA
  - Username Password Form - Required
  - 2FA - Required
    - OTP Flow - Alternative
      - OTP Form - Required
    - WebAuthn Authenticator - Alternative
```

This will cause the users that do not have any form of second factor authenticator registered to be asked to register the first alternative (OTP), since it is set to `Required` inside the `Alternative` flow, and there are no other valid alternatives that can pass. After the users log in, they can go to their account console and register a WebAuthn authenticator. After this, they can select the alternative on their next login.

The issue that you have mentioned appears to be linked to the fact that if users do not have any second factor authenticator registered, they are never asked to register one since both OTP and WebAuthn forms are set as alternatives. The subflow described above sets the OTP form as required. One note is that the OTP is not truely required, since users that have a WebAuthn registered and have removed their OTP from their account will only be asked for the WebAuthn token.

**Reply by claudioweiler** [reply]
It looks good @darius-m , thanks!
It's simpler.
If you post this flow as an answer, I will mark it as accepted.

**Reply by ashish1099** [reply]
works absolutely fine.
Thanks for writing this down

**Reply by dblas** [reply]
Hello,
Well done.
What if the user forgot his mandatory OTP and would like to use the alternative one that is to say Webauthn?

In my case, kc 21.0.2, your solution doesn't work that way.
When selecting "another way" in the passwordless context, a panel pops up to present alternatives: password OR webauthn.
But, in the case where alternatives are OTP Form OR Webauthn it's always OTP form that shows up.
Curious, isn't it? 
Is it a 21.0.2 bug or is there something missing in a *.ftl?

<img width="869" alt="Alternative between OTP and webauthn" src="https://user-images.githubusercontent.com/6929433/230026416-980d440b-fe22-47ff-98ce-1ee87322886e.png">

Thank you,
db

**Reply by darius-m** [reply]
Does the user have an OTP configured in their account, but not an WebAuthn token? If so, not having the option to select a different way to authenticate is the expected behavior, since in that case the second factor could be enforced (and would thus be expected). Being able to select WebAuthn as the alternative (for configuration) after the correct password is entered, without entering a valid OTP code, would be a security issue since 2FA would not be correctly enforced.

The reason why the flow I've mentioned before makes any sense is because for the first user login, they do not have any second factor configured, and allowing the user to login with a single factor is the only reasonable approach; however, after one of the second factors has been configured, using it becomes mandatory. The user can access their account console to configure alternative 2FA options at any point, if they are able to log in into their account.

**Reply by claudioweiler** [reply]
Hi @dblas!

> What if the user forgot his mandatory OTP and would like to use the alternative one that is to say Webauthn?

This flow is designed to work exactly this way. As @darius-m sad above, the user just had to have both 2FA methods configured, and select "another way", remembering that the main method can be changed by the user in his account page.

> In my case, kc 21.0.2, your solution doesn't work that way.
> When selecting "another way" in the passwordless context, a panel pops up to present alternatives: password OR webauthn.
> ...
> Curious, isn't it?
> Is it a 21.0.2 bug or is there something missing in a *.ftl?

Hum... I can't test this in KC 21.
But this flow isn't designed for **passwordless** authentication. Maybe I'm missing something here...

> But, in the case where alternatives are OTP Form OR Webauthn it's always OTP form that shows up.

As explained above user must have both methods configured and the user can change the primary method on his account page (old keycloak theme, AFAIK on new keycloak.v2 theme this isn't possible).
If "another way" do not appears to the user, them is possible that the user do not have webauthn configured.

**Reply by VonNao** [reply]
Isn't there anyway to make people choose which 2fa they want to user after login? I tried many flows but always the first mentioned flow will be present and users cant choose between otp and webauth on registration.

**Reply by darius-m** [reply]
At least for the flow mentioned above, users should have a "Try a different way" (or similar) button under the OTP input, but this only works if the user has the alternatives configured before attempting to log in. Users are not presented with the option to select an alternative if they don't configure multiple alternatives through the account console.

**Reply by darius-m** [reply]
As a note, presenting the option to configure an alternative during log in does not make sense, as this would bypass the MFA requirement. If the user does not have any MFA option configured, they will, of course, be presented with the option (and required) to register one during their first login. However, subsequent logins must enforce MFA, so alternatives must be configured through the account console (after logging in).

**Reply by VonNao** [reply]
I unterstand that. In general i would say it is a very special usecase. I configured the following flow 
![2024-08-05 11_37_11-Keycloak Administration Console](https://github.com/user-attachments/assets/275ecb9c-50db-4265-8fb9-037272761794)

Which results in the possibility of creating "Try Antoher way"

![2024-08-05 11_38_52-Sign in to](https://github.com/user-attachments/assets/4459fa67-bc35-49f0-af9b-0001a25aa2bc)

with the following options. The problem is that the naming is the result from the first subflow.

![2024-08-05 11_35_31-Sign in to](https://github.com/user-attachments/assets/257ba7c5-60b0-4b3f-8563-1a88cdfb43d6)

**Reply by darius-m** [reply]
I see the problem, I'm not sure why it's taking the information from the first step instead of the subflow. It's probably worth reporting a bug or requesting an enhancement via the Issues tab.

In your case, though, the flow is pretty common for most users. Since the "Username and password" as a required step, which is common between the subflows, you can create a single subflow that has "Username and password" as the first step, and then the MFA options as a subflow. You could use the flow I've described before (https://github.com/keycloak/keycloak/discussions/14988#discussioncomment-4011561). Unless you need something more specific, it's a good (enough) workaround. I do encourage requesting an enhancement if you have more specific requirements, however.

**Reply by VonNao** [reply]
I think i will create an issuer for my usecase. 

I have tested you idea.

![grafik](https://github.com/user-attachments/assets/efaef847-062d-413a-808b-2515692c6d5b)

The Problem is that 2fa skipped altogether. I Think that is because of the "Requried" User Password Auth"?

**Reply by darius-m** [reply]
The steps for MFA must be in a subflow that is required. If they are set as "Alternative" at the same level as the Username and Password form they will be skipped (if alternative flows are in the same subflow as a required step, they are considered disabled, and the authentication subflow is complete with just the username and password).

**Reply by VonNao** [reply]
Ah yeah. My bad. It is working as expected. Sadly now OTP will be the Default 2fa and webauth will not be available because the OTP Form dosent show "Try another way". 

May hope was that a menu like "Select Login method" would be created if the subflows are present with "alternative" set.

**Reply by claudioweiler** [reply]
@VonNao , use my suggested flow or @darius-m suggested flow and it will work as expected:

1. User has (mandatory) to config TOTP, if they don't have one configured yet
2. After first login, user can access account console and configure multiple webauthn devices
3. With TOTP configured, and at least one webauthn device configured, user will be presented with the option to chose another method as 2FA

Preference of 2FA method can be modified on admin console, per user. Account console lacks this functionality.

![image](https://github.com/user-attachments/assets/88cec653-d4f3-4939-857c-7901026496f0)

![image](https://github.com/user-attachments/assets/3403a5c0-7559-42b6-9a11-4651cbfcfa6c)

![image](https://github.com/user-attachments/assets/7a62dec3-c5b3-4b90-a72b-7b4c9125c46d)

**Reply by VonNao** [reply]
@claudioweiler Thanks for the detailed documentation. I created a similar flow. The only "problem", which may be how keycloak works, is that in our usecase it would be preferable if users could choose between registering TOTP or Webauth on the Subflow "Register OTP" before they even registered their first otp

**Reply by darius-m** [reply]
I've created an enhancement request - #31934; we'll see if it can be implemented.

**Reply by claudioweiler** [reply]
> ... The only "problem", which may be how keycloak works, is that in our usecase it would be preferable if users could choose between registering TOTP or Webauth on the Subflow "Register OTP" before they even registered their first otp

Right, got it.

My consideration on this is that TOTP is more simple and reliable to general users, so lower chances on support request to handle login problems.

My personal usecase, with webauthn as primary and TOTP as fallback works great for me. I can use webauth on most login scenarios, but when using an unknown device for webauthn, TOTP comes to rescue.

**Reply by VonNao** [reply]
@claudioweiler Yeah, totally understandable. Personally, I think most of our customers will choose OTP over Webauth anyway, but there are always the 5% who are a little bit special and need some extra treatment.

Anyway, in general it would be a nice addition, especially since tools like authelia support it out of the box.

For now, i guess we will stick to a flow like you mentioned above.

**Comment by jbman** [comment]
I filed a similar question. The difference is, that both second factors should be enforced to be configured: https://github.com/keycloak/keycloak/discussions/19548
Doesn't seem to be possible out-of-the-box.

**Reply by dblas** [reply]
Someone seems to have done [it](https://stackoverflow.com/questions/51980195/multiple-authentication-methods-for-a-user-in-keycloak) but he doesn't remember (or doesn't want to remember) how he did it.
db

**Reply by darius-m** [reply]
Configuring additional 2FA options can be enforced (at least partially) through required actions. When the user logs in, they have to go through some operations, like configuring OTP / WebAuthn, agreeing to terms and conditions, etc.. If an action should be assigned to a user when they are first created, default required actions can be used, where the action is automatically assigned to new users.

Making sure that the user has both 2FA factors configured is, however, not achievable through default required actions alone (as these only apply to new users, so if the user logs in and then removes the 2FA option through their account console, the required actions is not re-assigned). If you need this, you can probably use an SPI (i.e., a plugin) to check if the 2FA options are configured whenever the user logs in.

**Reply by jbman** [reply]
Required actions work, if every user should be enforced to use 2FA. But in case of step-up authentication, a client decides that some operation needs 2FA. The second factor (or two of them - one for recovery) needs to be setup only for users who trigger this client operation. 
I think the 2FA authenticators Webauthn, OTP Form and Recovery Codes should be configurable to enforce setup even if alternative in the flow.

**Reply by darius-m** [reply]
Yes, I also see the merit of having an option to enforce setting up multiple second factor methods. Maybe a simple checkbox in the flow interface, or the required actions tab, that checks the if the 2FA option is enabled at every login would suffice. However, I do not know how complex the implementation would be, since I would assume that the flow logic would have to check what 2FA methods there are, and how to not interfere with the other flows (i.e., allow the user to login using only their password if no 2FA option is available, but otherwise enforce 2FA). As far as I can tell, this is not an option out of the box, but it may be a good feature request.

**Comment by dblas** [comment]
Question was discussed, amongst other things, 4 years ago, [here](https://github.com/keycloak/keycloak-community/blob/main/design/multi-factor-admin-and-step-up.md).
Doesn't seem to have been implemented so far.
It's a pity,

db

**Comment by eggyknap** [comment]
This works nicely enough for me. I've configured OTP, WebAuthn, and Recovery Codes all as alternatives. I've required recovery code registration as a default required action in the realm, and in our case, administrators will need to assign OTP or WebAuthn as a required action for new users. This shouldn't be a problem in our case. Thanks for the ideas!

**Comment by PRD-1** [comment]
I want to develop a user authentication flow and need guidance on setting it up correctly. Currently, my process requires users to log in with a username and password, followed by a TAN (Transaction Authentication Number) sent to their email for further verification. I want to implement an optional feature where users can choose to register for a Web Authenticator instead of receiving a TAN. When a user opts for the Web Authenticator, the TAN verification step should automatically be skipped.

However, I'm encountering problems with the configuration:

When OTP and TAN are set as alternatives: The system fails to bypass the TAN step for users who are registered with the Web Authenticator, which it should ideally do. But it is also Bypassining the OTP step for registered Users with Web Authenticators.

When I set OTP as required and TAN as an alternative: If the user has neither method configured, the system either forces them to register upon login (which we want to avoid), or it fails to function correctly (i.e., does not authenticate them at all).

I need the Web Authenticator to be a voluntary, non-mandatory option that, when used, eliminates the need for TAN verification. How can I configure the system to:

Allow users to choose the Web Authenticator without making it mandatory?
Ensure that choosing the Web Authenticator bypasses the TAN step?
Avoid forcing users to register for any method as a prerequisite for login if they have not already set up an authentication method?

**Comment by andrewmeyer** [comment]
the workflow I was trying to accomplish was to "prefer" webauthn over OTP for users that have both set up.  the solution was to, as an admin, drag webauthn over OTP on the user's credential page - very unfortunate that this is not something the user can do, but I can probably count on one hand the number of users that would prefer this.

**Reply by claudioweiler** [reply]
If my memory isn't joking me, old v1 account console template had this. Maybe some great soul can help on guides to tweaking current template with a custom page or something like this.

**Comment by eggyknap** [comment]
Given a flow where 2fa is conditional upon having a 2fa credential
configured, what prevents a user from getting logged in, setting up their
2fa as required, but then deleting that 2fa credential through the account
console? Then their subsequent logins would not require 2fa, because they
no longer have a 2fa credential configured.

**Reply by darius-m** [reply]
I assume you wanted to reply to one of the threads? Regardless, the condition would indeed make the requirement of MFA not applicable. However, if the OTP is set to "Required" as the only step in an alternative subflow, when none of the alternatives can be applied, the "required" step is enforced.

Let's consider the following flow:
```
- Username and Password (required)
- MFA (required subflow)
  - TOTP (alternative subflow)
    - OTP Form (required)
  - WebAuthn Authenticator (alternative)
```

If the user has either OTP or WebAuthn configured, Keycloak will chose that particular step for authentication. If neither are set, Keycloak will enter the TOTP alternative flow, and since the "OTP Form" is set to required, it will ask the user to configure TOTP. Adding a condition in the "TOTP" flow would indeed disable both cases, and users would be unable to log in (the "MFA" parent subflow would not be able to succeed, so if it's set to required, the authentication flow would fail).

You can confirm whether this works as expected for your scenario.

**Reply by claudioweiler** [reply]
This discussion is about a "mandatory 2FA flow", so I will stick with this definition.

> Given a flow where 2fa is conditional upon having a 2fa credential configured, what prevents a user from getting logged in, ..., but then deleting that 2fa credential

2FA **method** is conditional, but 2FA **flow** is mandatory. If the user removes they 2FA methods, they will have to configure one again on next login.

**Comment by eggyknap** [comment]
I did want to reply to a thread, and messed the email up. Thanks for being
understanding, and thanks to all for responding. I thought I'd tried
something exactly like this and found a reason it wouldn't work for me.
I'll investigate that again. Thanks again.
